src/components/sidenav/sidenav.ts

```diff
@@ -0,0 +1,353 @@
+      if (this.opened_) {
+        if (this.openPromise_) {
+          this.openPromiseResolve_();
+        }
+        if (this.closePromise_) {
+          this.closePromiseReject_();
+        }
+
+        this.onOpen.emit(null);
+      } else {
+        if (this.closePromise_) {
+          this.closePromiseResolve_();
+        }
+        if (this.openPromise_) {
+          this.openPromiseReject_();
+        }
+
+        this.onClose.emit(null);
+      }
+
+      this.openPromise_ = null;
+      this.closePromise_ = null;
+    }
+  }
+
+  @HostBinding('class.md-sidenav-closing') private get isClosing_() {
+    return !this.opened_ && this.transition_;
+  }
+  @HostBinding('class.md-sidenav-opening') private get isOpening_() {
+    return this.opened_ && this.transition_;
+  }
+  @HostBinding('class.md-sidenav-closed') private get isClosed_() {
+    return !this.opened_ && !this.transition_;
+  }
+  @HostBinding('class.md-sidenav-opened') private get isOpened_() {
+    return this.opened_ && !this.transition_;
+  }
+  @HostBinding('class.md-sidenav-end') private get isEnd_() {
+    return this.align == 'end';
+  }
+  @HostBinding('class.md-sidenav-side') private get modeSide_() {
+    return this.mode == 'side';
+  }
+  @HostBinding('class.md-sidenav-over') private get modeOver_() {
+    return this.mode == 'over';
+  }
+  @HostBinding('class.md-sidenav-push') private get modePush_() {
+    return this.mode == 'push';
+  }
+
+  /**
+   * This is public because we need it from MdSidenavLayout, but it's undocumented and should
+   * not be used outside.
+   * @returns {number}
+   * @private
+   */
+  public get width_() {
+    if (this.elementRef_.nativeElement) {
+      return this.elementRef_.nativeElement.offsetWidth;
+    }
+    return 0;
+  }
+
+  private transition_: boolean = false;
+  private openPromise_: Promise<void>;
+  private openPromiseResolve_: () => void;
+  private openPromiseReject_: () => void;
+  private closePromise_: Promise<void>;
+  private closePromiseResolve_: () => void;
+  private closePromiseReject_: () => void;
+}
+
+
+
+/**
+ * <md-sidenav-layout> component.
+ */
+@Component({
+  selector: 'md-sidenav-layout',
+  directives: [MdSidenav],
+  templateUrl: './components/sidenav/sidenav.html',
+  styleUrls: ['./components/sidenav/sidenav.css'],
+})
+export class MdSidenavLayout implements AfterContentInit {
+  @ContentChildren(MdSidenav) private drawers_: QueryList<MdSidenav>;
+
+  get start() { return this.start_; }
+  get end() { return this.end_; }
+
+  constructor(@Optional() @Host() private dir_: Dir) {
+    if (dir_) {
+      dir_.dirChange.subscribe(() => this.validateDrawers_());
+    }
+  }
+
+  ngAfterContentInit() {
+    // On changes, assert on consistency.
+    this.drawers_.changes.subscribe(() => this.validateDrawers_());
+    this.validateDrawers_();
+  }
+
+
+
+  private start_: MdSidenav;
+  private end_: MdSidenav;
+  private right_: MdSidenav;
+  private left_: MdSidenav;
```

Comment:
Why are there `left_` and `right_` properties? If they're just aliases, I think that's confusing and prone to errors.